geps/gep-3155/index.md

```diff
@@ -0,0 +1,205 @@
+  //
+  // References to a resource in different namespace are invalid UNLESS there
+  // is a ReferenceGrant in the target namespace that allows the certificate
+  // to be attached. If a ReferenceGrant does not allow this reference, the
+  // "ResolvedRefs" condition MUST be set to False for this listener with the
+  // "RefNotPermitted" reason.
+  //
+  // CertificateRef can reference to standard Kubernetes resources, i.e.
+  // Secret, or implementation-specific custom resources.
+  //
+  // If CertificateRef is not specified, no client certificate is presented when
+  // connecting to the service.
+  CertificateRef SecretObjectReference `json:"certificateRef,omitempty"`
+}
+```
+
+#### Limitations
+
+Configuring client certificate on the service level may result in multiple Gateways
+from different vendors sharing the same identity when connecting to a single service.
+
+The problem could be partially mitigated by allowing shared policies to selectively 
+target specific Gateway instances, GatewayClasses or gateway namespaces.
+
+### SANs on BackendTLSPolicy
+
+This change enables the certificate to have a different identity than the SNI
+(both are currently tied to the hostname field). This is particularly useful
+when using SPIFFE, which relies on URI Subject Names which are not valid SNIs 
+as per https://www.rfc-editor.org/rfc/rfc6066.html#section-3.
+
+In such case either connection properties or an arbitrary SNI, like cluster-local 
+service name could be used for certificate selection, while the identity validation
+will be done based on SubjectAltNames field.
+
+**1. Add a new `SubjectAltNames` field to `BackendTLSPolicyValidation`**
+
+```go
+type BackendTLSPolicyValidation struct {
+  // SubjectAltNames contains one or more Subject Alternative Names.
+  // When specified, at least one of certificate's Subject Alternate Names MUST 
+  // match at least one of the specified SubjectAltNames.
+  // +kubebuilder:validation:MaxItems=5
+  SubjectAltNames []SubjectAltName `json:"subjectAltNames,omitempty"`
+}
+
+type SubjectAltName struct {
+  // Hostname contains Subject Alternative Name specified in DNS name format. It is
+  // mutually exclusive with Hostname. At least one of the fields must be specified.
+  Hostname v1.PreciseHostname `json:"hostname"`
+
+  // URI contains Subject Alternative Name specified in URI format. It is mutually
+  // exclusive with Hostname. At least one of the fields must be specified.
+  URI string `json:"uri,omitempty"`
+}
+```
+
+**2. Modify Spec for `BackendTLSPolicyValidation` `Hostname`**
+
+Before:
+```go
+	// 2. Hostname MUST be used for authentication and MUST match the certificate
+	//    served by the matching backend.
+```
+
+After:
+```go
+	// 2. Only when SubjectAltNames is not specified, Hostname MUST be used for 
+    //    authentication and MUST match the certificate served by the matching
+    //    backend.
+    // 3. If HostName is unspecified, SNI header will be omitted.
```

Comment:
If there is no SNI, how can the server choose what certificate to serve? This implies but does not state that the server must only have a single serving certificate to choose from. If that's an assumption, I guess that's okay, but we need to be clear about the contract here.

Maybe we need to add something to some docstrings somewhere to say something like "BackendTLSPolicy expects that backends covered by it SHOULD have only one serving keypair to choose from, as the SNI field may not always be set" or something.

Reply:
I think that the common convention is to provide the "default" certificate in such case.

I think I would rather use language that the Hostname SHOULD be sad - otherwise server will have hard time selecting the correct certificate, but I wanted to allow unset for compatibility corner cases.

But again, if there is a consensus we don't need to support this case -> I am more than happy to drop this from the spec.

Reply:
Agreed, this seems to be pretty niche, let's remove it.